File: languages/csharp/LibLogit/Logit.cs
using System.Text.Json;

namespace LibLogit;

public enum Level
{
    Trace = 0,
    Debug = 1,
    Info = 2,
    Warn = 3,
    Error = 4,
    Fatal = 5
}

public sealed class Logit
{
    private string? _localPath;
    private string? _remotePath;
    private readonly List<string> _sinks = new() { "console" };

    public string Name { get; set; } = "default";
    public Level Level { get; set; } = Level.Info;
    public bool Timestamp { get; set; } = true;
    public bool TagLevel { get; set; } = true;
    public string Format { get; set; } = "text";
    public Dictionary<string, object?> Metadata { get; set; } = new();
    public bool Enabled { get; set; } = true;

    public string? LocalPath
    {
        get => _localPath;
        set
        {
            _localPath = value;
            SetSink("file", !string.IsNullOrWhiteSpace(value));
        }
    }

    public string? RemotePath
    {
        get => _remotePath;
        set
        {
            _remotePath = value;
            SetSink("network", !string.IsNullOrWhiteSpace(value));
        }
    }

    public IReadOnlyList<string> Sinks
    {
        get => _sinks;
        set
        {
            _sinks.Clear();
            foreach (var sink in value)
            {
                AddSink(sink);
            }
        }
    }

    public Logit()
    {
    }

    public Logit(LogitOptions options)
    {
        Name = string.IsNullOrWhiteSpace(options.Name) ? "default" : options.Name;
        Level = ParseLevel(options.Level ?? "info");
        Timestamp = options.Timestamp ?? true;
        TagLevel = options.TagLevel ?? true;
        Format = options.Format ?? "text";
        Metadata = options.Metadata ?? new Dictionary<string, object?>();
        Enabled = options.Enabled ?? true;
        LocalPath = options.LocalPath ?? options.Path;
        RemotePath = options.RemotePath ?? options.NetworkPath;
        if (options.Sinks is not null)
        {
            Sinks = options.Sinks;
        }
    }

    public LogBuilder At(Level level) => new(this, level);

    public LogBuilder At(string level) => At(ParseLevel(level));

    public void Log(Level level, object? message)
    {
        if (!Enabled || level < Level)
        {
            return;
        }

        var rendered = Render(level, Stringify(message));
        if (_sinks.Contains("console"))
        {
            Console.Error.WriteLine(rendered);
        }
        if (_sinks.Contains("file") && !string.IsNullOrWhiteSpace(LocalPath))
        {
            AppendLine(LocalPath, rendered);
        }
        if (_sinks.Contains("network") && !string.IsNullOrWhiteSpace(RemotePath) && !LooksLikeSocket(RemotePath))
        {
            AppendLine(RemotePath, rendered);
        }
    }

    public void Log(string level, object? message) => Log(ParseLevel(level), message);

    private string Render(Level level, string message)
    {
        if (string.Equals(Format, "json", StringComparison.OrdinalIgnoreCase))
        {
            var payload = new SortedDictionary<string, object?>
            {
                ["level"] = CanonicalLevel(level),
                ["logger"] = Name,
                ["message"] = message
            };
            if (Metadata.Count > 0)
            {
                payload["metadata"] = Metadata;
            }
            if (Timestamp)
            {
                payload["timestamp"] = DateTimeOffset.UtcNow.ToString("O");
            }
            return JsonSerializer.Serialize(payload);
        }

        var parts = new List<string>();
        if (Timestamp)
        {
            parts.Add(DateTimeOffset.Now.ToString("yyyy-MM-ddTHH:mm:sszzz"));
        }
        if (TagLevel)
        {
            parts.Add(RenderedLevel(level));
        }
        parts.Add(message);
        return string.Join(" ", parts);
    }

    public static Dictionary<string, Logit> LoadLogits(string configPath)
    {
        using var document = JsonDocument.Parse(File.ReadAllText(configPath));
        var root = document.RootElement;
        if (!root.TryGetProperty("logits", out var logitsElement))
        {
            var level = "info";
            var tagLevel = true;
            if (root.TryGetProperty("level", out var levelElement))
            {
                if (levelElement.ValueKind == JsonValueKind.String)
                {
                    level = levelElement.GetString() ?? "info";
                }
                else
                {
                    level = levelElement.GetProperty("threshold").GetString() ?? "info";
                    tagLevel = !levelElement.TryGetProperty("tag", out var tagElement) || tagElement.GetBoolean();
                }
            }
            return new Dictionary<string, Logit>
            {
                ["default"] = new(new LogitOptions
                {
                    Name = "default",
                    Level = level,
                    TagLevel = tagLevel,
                    Timestamp = !root.TryGetProperty("timestamp", out var timestamp) || timestamp.GetBoolean(),
                    Path = root.TryGetProperty("file_location", out var file) && file.ValueKind == JsonValueKind.String ? file.GetString() : null,
                    RemotePath = root.TryGetProperty("network_file_location", out var remote) && remote.ValueKind == JsonValueKind.String ? remote.GetString() : null
                })
            };
        }

        var defaults = root.TryGetProperty("defaults", out var defaultsElement)
            ? ReadOptions(defaultsElement, "default")
            : new LogitOptions();

        var result = new Dictionary<string, Logit>();
        foreach (var logitProperty in logitsElement.EnumerateObject())
        {
            var options = Merge(defaults, ReadOptions(logitProperty.Value, logitProperty.Name));
            options.Name ??= logitProperty.Name;
            result[options.Name] = new Logit(options);
        }
        return result;
    }

    private static LogitOptions ReadOptions(JsonElement element, string name)
    {
        var options = new LogitOptions { Name = name };
        if (element.TryGetProperty("name", out var nameElement))
        {
            options.Name = nameElement.GetString();
        }
        if (element.TryGetProperty("level", out var levelElement))
        {
            options.Level = levelElement.GetString();
        }
        if (element.TryGetProperty("path", out var pathElement))
        {
            options.Path = pathElement.ValueKind == JsonValueKind.String ? pathElement.GetString() : null;
        }
        if (element.TryGetProperty("localPath", out var localPathElement))
        {
            options.LocalPath = localPathElement.ValueKind == JsonValueKind.String ? localPathElement.GetString() : null;
        }
        if (element.TryGetProperty("remotePath", out var remotePathElement))
        {
            options.RemotePath = remotePathElement.ValueKind == JsonValueKind.String ? remotePathElement.GetString() : null;
        }
        if (element.TryGetProperty("network_path", out var networkPathElement))
        {
            options.NetworkPath = networkPathElement.ValueKind == JsonValueKind.String ? networkPathElement.GetString() : null;
        }
        if (element.TryGetProperty("timestamp", out var timestampElement))
        {
            options.Timestamp = timestampElement.GetBoolean();
        }
        if (element.TryGetProperty("tag_level", out var tagLevelElement))
        {
            options.TagLevel = tagLevelElement.GetBoolean();
        }
        if (element.TryGetProperty("format", out var formatElement))
        {
            options.Format = formatElement.GetString();
        }
        if (element.TryGetProperty("sinks", out var sinksElement))
        {
            options.Sinks = sinksElement.EnumerateArray().Select(item => item.GetString() ?? "").Where(item => item.Length > 0).ToArray();
        }
        if (element.TryGetProperty("metadata", out var metadataElement))
        {
            options.Metadata = JsonSerializer.Deserialize<Dictionary<string, object?>>(metadataElement.GetRawText()) ?? new();
        }
        return options;
    }

    private static LogitOptions Merge(LogitOptions defaults, LogitOptions current)
    {
        return new LogitOptions
        {
            Name = current.Name ?? defaults.Name,
            Path = current.Path ?? defaults.Path,
            LocalPath = current.LocalPath ?? defaults.LocalPath,
            RemotePath = current.RemotePath ?? defaults.RemotePath,
            NetworkPath = current.NetworkPath ?? defaults.NetworkPath,
            Level = current.Level ?? defaults.Level,
            Enabled = current.Enabled ?? defaults.Enabled,
            Sinks = current.Sinks ?? defaults.Sinks,
            Timestamp = current.Timestamp ?? defaults.Timestamp,
            TagLevel = current.TagLevel ?? defaults.TagLevel,
            Format = current.Format ?? defaults.Format,
            Metadata = current.Metadata ?? defaults.Metadata
        };
    }

    private void SetSink(string sink, bool enabled)
    {
        if (enabled)
        {
            AddSink(sink);
            return;
        }
        _sinks.Remove(sink);
    }

    private void AddSink(string sink)
    {
        var normalized = sink.ToLowerInvariant();
        if (normalized is not ("console" or "file" or "network"))
        {
            throw new ArgumentException($"Unknown sink: {sink}");
        }
        if (!_sinks.Contains(normalized))
        {
            _sinks.Add(normalized);
        }
    }

    private static void AppendLine(string target, string line)
    {
        var parent = Path.GetDirectoryName(target);
        if (!string.IsNullOrEmpty(parent))
        {
            Directory.CreateDirectory(parent);
        }
        File.AppendAllText(target, line + Environment.NewLine);
    }

    private static bool LooksLikeSocket(string target) =>
        target.StartsWith("tcp://", StringComparison.OrdinalIgnoreCase)
        || target.StartsWith("udp://", StringComparison.OrdinalIgnoreCase);

    private static string Stringify(object? value) =>
        value switch
        {
            null => "null",
            string text => text,
            _ => JsonSerializer.Serialize(value)
        };

    public static Level ParseLevel(string level)
    {
        return level.Trim().ToLowerInvariant() switch
        {
            "trace" => Level.Trace,
            "debug" => Level.Debug,
            "info" => Level.Info,
            "warn" or "warning" => Level.Warn,
            "error" => Level.Error,
            "fatal" => Level.Fatal,
            _ => throw new ArgumentException($"Unknown level: {level}")
        };
    }

    private static string CanonicalLevel(Level level) => level switch
    {
        Level.Trace => "trace",
        Level.Debug => "debug",
        Level.Info => "info",
        Level.Warn => "warn",
        Level.Error => "error",
        Level.Fatal => "fatal",
        _ => "info"
    };

    private static string RenderedLevel(Level level) => level switch
    {
        Level.Trace => "TRACE",
        Level.Debug => "DEBUG",
        Level.Info => "INFO",
        Level.Warn => "WARNING",
        Level.Error => "ERROR",
        Level.Fatal => "CRITICAL",
        _ => "INFO"
    };
}

public sealed class LogBuilder
{
    private readonly Logit _logit;
    private readonly Level _level;
    private readonly List<string> _fragments = new();
    private bool _committed;

    internal LogBuilder(Logit logit, Level level)
    {
        _logit = logit;
        _level = level;
    }

    public LogBuilder Append(object? value)
    {
        if (!_committed)
        {
            _fragments.Add(value?.ToString() ?? "null");
        }
        return this;
    }

    public LogBuilder Commit()
    {
        if (!_committed && _fragments.Count > 0)
        {
            _committed = true;
            _logit.Log(_level, string.Concat(_fragments));
        }
        return this;
    }
}

public sealed class LogitOptions
{
    public string? Name { get; set; }
    public string? Path { get; set; }
    public string? LocalPath { get; set; }
    public string? RemotePath { get; set; }
    public string? NetworkPath { get; set; }
    public string? Level { get; set; }
    public bool? Enabled { get; set; }
    public IReadOnlyList<string>? Sinks { get; set; }
    public bool? Timestamp { get; set; }
    public bool? TagLevel { get; set; }
    public string? Format { get; set; }
    public Dictionary<string, object?>? Metadata { get; set; }
}

File: languages/go/liblogit/logit.go
package liblogit

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

type Level int

const (
	Trace Level = iota
	Debug
	Info
	Warn
	Error
	Fatal
)

type Logit struct {
	Name       string
	LocalPath string
	RemotePath string
	Level      Level
	Enabled    bool
	Sinks      []string
	Timestamp  bool
	TagLevel   bool
	Format     string
	Metadata   map[string]any
}

type Options struct {
	Name       string         `json:"name"`
	Path       string         `json:"path"`
	LocalPath  string         `json:"localPath"`
	RemotePath string         `json:"remotePath"`
	NetworkPath string        `json:"network_path"`
	Level      string         `json:"level"`
	Enabled    *bool          `json:"enabled"`
	Sinks      []string       `json:"sinks"`
	Timestamp  *bool          `json:"timestamp"`
	TagLevel   *bool          `json:"tag_level"`
	Format     string         `json:"format"`
	Metadata   map[string]any `json:"metadata"`
}

func New() *Logit {
	return &Logit{
		Name:      "default",
		Level:     Info,
		Enabled:   true,
		Sinks:     []string{"console"},
		Timestamp: true,
		TagLevel:  true,
		Format:    "text",
		Metadata:  map[string]any{},
	}
}

func FromOptions(options Options) (*Logit, error) {
	logit := New()
	if options.Name != "" {
		logit.Name = options.Name
	}
	if options.Level != "" {
		level, err := ParseLevel(options.Level)
		if err != nil {
			return nil, err
		}
		logit.Level = level
	}
	if options.Enabled != nil {
		logit.Enabled = *options.Enabled
	}
	if options.Timestamp != nil {
		logit.Timestamp = *options.Timestamp
	}
	if options.TagLevel != nil {
		logit.TagLevel = *options.TagLevel
	}
	if options.Format != "" {
		logit.Format = options.Format
	}
	if options.Metadata != nil {
		logit.Metadata = options.Metadata
	}
	localPath := first(options.LocalPath, options.Path)
	if localPath != "" {
		logit.SetLocalPath(localPath)
	}
	remotePath := first(options.RemotePath, options.NetworkPath)
	if remotePath != "" {
		logit.SetRemotePath(remotePath)
	}
	if options.Sinks != nil {
		logit.Sinks = normalizeSinks(options.Sinks)
	}
	return logit, nil
}

func (l *Logit) SetLocalPath(value string) {
	l.LocalPath = value
	l.setSink("file", value != "")
}

func (l *Logit) SetRemotePath(value string) {
	l.RemotePath = value
	l.setSink("network", value != "")
}

func (l *Logit) At(level Level) *Builder {
	return &Builder{logit: l, level: level}
}

func (l *Logit) AtString(level string) (*Builder, error) {
	parsed, err := ParseLevel(level)
	if err != nil {
		return nil, err
	}
	return l.At(parsed), nil
}

func (l *Logit) Log(level Level, message any) error {
	if !l.Enabled || level < l.Level {
		return nil
	}
	rendered, err := l.render(level, stringify(message))
	if err != nil {
		return err
	}
	if contains(l.Sinks, "console") {
		fmt.Fprintln(os.Stderr, rendered)
	}
	if contains(l.Sinks, "file") && l.LocalPath != "" {
		if err := appendLine(l.LocalPath, rendered); err != nil {
			return err
		}
	}
	if contains(l.Sinks, "network") && l.RemotePath != "" && !looksLikeSocket(l.RemotePath) {
		if err := appendLine(l.RemotePath, rendered); err != nil {
			return err
		}
	}
	return nil
}

func (l *Logit) render(level Level, message string) (string, error) {
	if strings.EqualFold(l.Format, "json") {
		event := map[string]any{
			"level":   canonicalLevel(level),
			"logger":  l.Name,
			"message": message,
		}
		if len(l.Metadata) > 0 {
			event["metadata"] = l.Metadata
		}
		if l.Timestamp {
			event["timestamp"] = time.Now().UTC().Format(time.RFC3339Nano)
		}
		bytes, err := json.Marshal(sortedMap(event))
		return string(bytes), err
	}

	parts := []string{}
	if l.Timestamp {
		parts = append(parts, time.Now().Format("2006-01-02T15:04:05-07:00"))
	}
	if l.TagLevel {
		parts = append(parts, renderedLevel(level))
	}
	parts = append(parts, message)
	return strings.Join(parts, " "), nil
}

func LoadLogits(configPath string) (map[string]*Logit, error) {
	bytes, err := os.ReadFile(configPath)
	if err != nil {
		return nil, err
	}
	var root map[string]json.RawMessage
	if err := json.Unmarshal(bytes, &root); err != nil {
		return nil, err
	}
	if _, ok := root["logits"]; !ok {
		var legacy struct {
			Level any `json:"level"`
			Timestamp *bool `json:"timestamp"`
			FileLocation string `json:"file_location"`
			NetworkFileLocation string `json:"network_file_location"`
		}
		if err := json.Unmarshal(bytes, &legacy); err != nil {
			return nil, err
		}
		options := Options{Name: "default", Timestamp: legacy.Timestamp, Path: legacy.FileLocation, RemotePath: legacy.NetworkFileLocation}
		switch level := legacy.Level.(type) {
		case string:
			options.Level = level
		case map[string]any:
			if threshold, ok := level["threshold"].(string); ok {
				options.Level = threshold
			}
		}
		logit, err := FromOptions(options)
		if err != nil {
			return nil, err
		}
		return map[string]*Logit{"default": logit}, nil
	}

	var config struct {
		Defaults Options            `json:"defaults"`
		Logits   map[string]Options `json:"logits"`
	}
	if err := json.Unmarshal(bytes, &config); err != nil {
		return nil, err
	}
	result := map[string]*Logit{}
	for name, options := range config.Logits {
		merged := mergeOptions(config.Defaults, options)
		if merged.Name == "" {
			merged.Name = name
		}
		logit, err := FromOptions(merged)
		if err != nil {
			return nil, err
		}
		result[logit.Name] = logit
	}
	return result, nil
}

type Builder struct {
	logit *Logit
	level Level
	fragments []string
	committed bool
}

func (b *Builder) Append(value any) *Builder {
	if !b.committed {
		b.fragments = append(b.fragments, stringify(value))
	}
	return b
}

func (b *Builder) Commit() error {
	if b.committed || len(b.fragments) == 0 {
		return nil
	}
	b.committed = true
	return b.logit.Log(b.level, strings.Join(b.fragments, ""))
}

func ParseLevel(value string) (Level, error) {
	switch strings.ToLower(strings.TrimSpace(value)) {
	case "trace":
		return Trace, nil
	case "debug":
		return Debug, nil
	case "info":
		return Info, nil
	case "warn", "warning":
		return Warn, nil
	case "error":
		return Error, nil
	case "fatal":
		return Fatal, nil
	default:
		return Info, fmt.Errorf("unknown level: %s", value)
	}
}

func (l *Logit) setSink(sink string, enabled bool) {
	if enabled && !contains(l.Sinks, sink) {
		l.Sinks = append(l.Sinks, sink)
	}
	if !enabled {
		filtered := []string{}
		for _, existing := range l.Sinks {
			if existing != sink {
				filtered = append(filtered, existing)
			}
		}
		l.Sinks = filtered
	}
}

func mergeOptions(defaults Options, current Options) Options {
	merged := defaults
	if current.Name != "" { merged.Name = current.Name }
	if current.Path != "" { merged.Path = current.Path }
	if current.LocalPath != "" { merged.LocalPath = current.LocalPath }
	if current.RemotePath != "" { merged.RemotePath = current.RemotePath }
	if current.NetworkPath != "" { merged.NetworkPath = current.NetworkPath }
	if current.Level != "" { merged.Level = current.Level }
	if current.Enabled != nil { merged.Enabled = current.Enabled }
	if current.Sinks != nil { merged.Sinks = current.Sinks }
	if current.Timestamp != nil { merged.Timestamp = current.Timestamp }
	if current.TagLevel != nil { merged.TagLevel = current.TagLevel }
	if current.Format != "" { merged.Format = current.Format }
	if current.Metadata != nil { merged.Metadata = current.Metadata }
	return merged
}

func normalizeSinks(values []string) []string {
	result := []string{}
	for _, value := range values {
		normalized := strings.ToLower(value)
		if !contains(result, normalized) {
			result = append(result, normalized)
		}
	}
	return result
}

func appendLine(target string, line string) error {
	if parent := filepath.Dir(target); parent != "." && parent != "" {
		if err := os.MkdirAll(parent, 0o755); err != nil {
			return err
		}
	}
	file, err := os.OpenFile(target, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o644)
	if err != nil {
		return err
	}
	defer file.Close()
	_, err = fmt.Fprintln(file, line)
	return err
}

func stringify(value any) string {
	switch typed := value.(type) {
	case string:
		return typed
	default:
		bytes, err := json.Marshal(value)
		if err != nil {
			return fmt.Sprint(value)
		}
		return string(bytes)
	}
}

func sortedMap(value map[string]any) map[string]any {
	keys := make([]string, 0, len(value))
	for key := range value {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	result := map[string]any{}
	for _, key := range keys {
		result[key] = value[key]
	}
	return result
}

func first(values ...string) string {
	for _, value := range values {
		if value != "" {
			return value
		}
	}
	return ""
}

func contains(values []string, target string) bool {
	for _, value := range values {
		if value == target {
			return true
		}
	}
	return false
}

func looksLikeSocket(value string) bool {
	lower := strings.ToLower(value)
	return strings.HasPrefix(lower, "tcp://") || strings.HasPrefix(lower, "udp://")
}

func canonicalLevel(value Level) string {
	switch value {
	case Trace: return "trace"
	case Debug: return "debug"
	case Info: return "info"
	case Warn: return "warn"
	case Error: return "error"
	case Fatal: return "fatal"
	default: return "info"
	}
}

func renderedLevel(value Level) string {
	switch value {
	case Trace: return "TRACE"
	case Debug: return "DEBUG"
	case Info: return "INFO"
	case Warn: return "WARNING"
	case Error: return "ERROR"
	case Fatal: return "CRITICAL"
	default: return "INFO"
	}
}
